src/System.Private.CoreLib/shared/System/Text/UTF8Encoding.cs

```diff
@@ -377,2120 +539,245 @@ public override unsafe int GetCharCount(ReadOnlySpan<byte> bytes)
+        private protected sealed override unsafe int GetCharsFast(byte* pBytes, int bytesLength, char* pChars, int charsLength, out int bytesConsumed)
         {
-            return (int)(((uint)((byte*)a - (byte*)b)) >> 1);
-        }
+            // We don't care about the exact OperationStatus value returned by the workhorse routine; we only
+            // care if the workhorse was able to consume the entire input payload. If we're unable to do so,
+            // we'll handle the remainder in the fallback routine.
 
-        // byte* flavor just for parity
-        private static unsafe int PtrDiff(byte* a, byte* b)
-        {
-            return (int)(a - b);
-        }
+            Utf8Utility.TranscodeToUtf16(pBytes, bytesLength, pChars, charsLength, out byte* pInputBufferRemaining, out char* pOutputBufferRemaining);
 
-        private static bool InRange(int ch, int start, int end)
-        {
-            return (uint)(ch - start) <= (uint)(end - start);
+            bytesConsumed = (int)(pInputBufferRemaining - pBytes);
+            return (int)(pOutputBufferRemaining - pChars);
         }
 
-        // Our workhorse
-        // Note:  We ignore mismatched surrogates, unless the exception flag is set in which case we throw
-        internal sealed override unsafe int GetBytes(
-            char* chars, int charCount, byte* bytes, int byteCount, EncoderNLS baseEncoder)
+        private protected sealed override unsafe int GetCharsWithFallback(ReadOnlySpan<byte> bytes, int originalBytesLength, Span<char> chars, int originalCharsLength, DecoderNLS decoder)
         {
-            Debug.Assert(chars != null, "[UTF8Encoding.GetBytes]chars!=null");
-            Debug.Assert(byteCount >= 0, "[UTF8Encoding.GetBytes]byteCount >=0");
-            Debug.Assert(charCount >= 0, "[UTF8Encoding.GetBytes]charCount >=0");
-            Debug.Assert(bytes != null, "[UTF8Encoding.GetBytes]bytes!=null");
-
-            UTF8Encoder encoder = null;
-
-            // For fallback we may need a fallback buffer.
-            // We wait to initialize it though in case we don't have any broken input unicode
-            EncoderFallbackBuffer fallbackBuffer = null;
-            char* pSrcForFallback;
-
-            char* pSrc = chars;
-            byte* pTarget = bytes;
-
-            char* pEnd = pSrc + charCount;
-            byte* pAllocatedBufferEnd = pTarget + byteCount;
-
-            int ch = 0;
-
-            // assume that JIT will en-register pSrc, pTarget and ch
+            // We special-case DecoderReplacementFallback if it's telling us to write a single U+FFFD char,
+            // since we believe this to be relatively common and we can handle it more efficiently than
+            // the base implementation.
 
-            if (baseEncoder != null)
+            if (((decoder is null) ? this.DecoderFallback : decoder.Fallback) is DecoderReplacementFallback replacementFallback
```

Comment:
when would this not be a `DecoderReplacementFallback`?

Reply:
Somebody can replace the fallback mechanism. For example: `new UTF8Encoding(..., throwOnInvalidBytes: true)` or `Encoding.Create("utf-8", ..., new MyCustomFallback())`.